### PR TITLE
[IMP] mail: do not loose message on ESC or no connectivity

### PR DESCRIPTION
**Pull request by alexkuhn**
Before these commits, we could lost a message with the following cases:

    - Accidental press ESC on full composer modal
    - Trying to send a message when we lost connection

These commits solve both issues.

Also, when closing a chat window with non-empty content in the composer, the user must confirm the close action in a dialog, in order to prevent loosing the message.

Task-ID 32923

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Elkasitu** [comment]
Vim users rejoice! 🤣

File: addons/mail/models/mail_activity.py [hunk 1]
```diff
@@ -229,7 +229,12 @@ def unlink(self):
     def action_done(self):
         """ Wrapper without feedback because web button add context as
         parameter, therefore setting context to feedback """
-        return self.action_feedback()
+        # return self.action_feedback()
```
Comment (adr-odoo) [inline addons/mail/models/mail_activity.py]: remove

File: addons/mail/static/src/js/draft_mixin.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    /**
+     * Get the draft content that has been saved (from full-composer)
+     *
+     * @private
+     * @return {string} draft message
+     */
+    _getDraft: function () {
+        return this._draft;
+    },
+    _isClosedOnSent: function (reason) {
```
Comment (adr-odoo) [inline addons/mail/static/src/js/draft_mixin.js]: ?

File: addons/mail/static/src/js/chatter.js [hunk 3]
```diff
@@ -315,6 +316,27 @@ var Chatter = Widget.extend({
...
+     *
+     * Note that we do not simply call the handler functions, because
+     * we would like to have UI changes automatically handled.
+     *
+     * @private
+     * @param {OdooEvent} ev
+     * @param {Object} ev.data
+     * @param {boolean} [ev.data.isLog=false]
+     */
+    _onOpenComposer: function (ev) {
```
Comment (adr-odoo) [inline addons/mail/static/src/js/chatter.js]: use _openComposer

File: addons/mail/static/src/js/draft_mixin.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
+odoo.define('mail.DraftMixin', function (require) {
+"use strict";
+
+var DraftMixin = {
+    /**
+     * @private
+     */
+    _autosaveDraft: function () {
+        var self = this;
+        var $editor = $('.modal .o_act_window .note-editable');
```
Comment (adr-odoo) [inline addons/mail/static/src/js/draft_mixin.js]: avoid global selector

File: addons/web/static/src/js/core/dialog.js [hunk 5]
```diff
@@ -165,12 +165,19 @@ var Dialog = Widget.extend({
...
+     * @param {string} [params.close_reason] a reason for closing the dialog.
+     * @param {boolean} [params.silent]
+     */
+    destroy: function (params) {
         // Need to trigger before real destroy but if 'closed' handler destroys
         // the widget again, we want to avoid infinite recursion
         if (!this.__closed) {
             this.__closed = true;
-            this.trigger("closed", arg);
+            this.trigger("closed", params);
```
Comment (aab-odoo) [inline addons/web/static/src/js/core/dialog.js]: no diff here

File: addons/web/static/src/js/chrome/action_manager.js [hunk 6]
```diff
@@ -275,20 +275,24 @@ var ActionManager = Widget.extend({
         }
     },
     /**
+     *
```
Comment (aab-odoo) [inline addons/web/static/src/js/chrome/action_manager.js]: no :)

File: addons/web/static/src/js/chrome/action_manager.js [hunk 7]
```diff
@@ -502,12 +502,12 @@ var ActionManager = Widget.extend({
...
     _executeCloseAction: function (action, options) {
-        var closeReason = action.close_reason;
+        var params = action || {};
         if (!this.currentDialogController) {
-            options.on_close(closeReason);
+            options.on_close(params.close_reason);
         }
 
-        this._closeDialog(closeReason);
+        this._closeDialog(params);
```
Comment (aab-odoo) [inline addons/web/static/src/js/chrome/action_manager.js]:
```
    _executeCloseAction: function (action, options) {
        if (!this.currentDialogController) {
            options.on_close(action.close_reason);
        }

        this._closeDialog({closeReason: action.close_reason});

        // display some effect (like rainbowman) on appropriate actions
        if (action.effect) {
            this.trigger_up('show_effect', action.effect);
        }

        return $.when();
    },
```

**Comment by pedrobaeza** [comment]
@alexkuhn has this reached v12 in any form or is it still pending?

**Comment by alexkuhn** [comment]
@pedrobaeza None of the features in this PR have reached v12.

There are some technical limitations that prevented us to implement this task, like the one mentioned in my previous comment.

The commits on connectivity loss could have been merged on their own, though...

**Comment by pedrobaeza** [comment]
@alexkuhn thanks for the followup. That's so sorry. I hope you will find another way. About the connectivity loss, it would be great if you rescue it, at least for v13.

**Comment by pedrobaeza** [comment]
This PR got stalled and with conflicts. Are some of the features already merged in v13/master? Should this be closed?

**Comment by alexkuhn** [comment]
Closing this PR, because this is outdated.
Currently working on significant changes for master, so that we can deal with these issues the right way.

**Comment by pedrobaeza** [comment]
Good to hear that! Eager to see them. Please ping me in the PR for checking.

**Comment by alexkuhn** [comment]
This requires writing discuss with OWL, which is still in draft: https://github.com/odoo/odoo/pull/39023

**Comment by pedrobaeza** [comment]
Wow, I see. That's too much to digest for me for now, but for sure you'll make a great work. When that gets merged, I will do some testing and hopefully provide some feedback.

Happy New Year!

**Comment by alexkuhn** [comment]
Really hope you will enjoy the improvements.

Happy New Year to you too!

**Pull request closed without merging**